src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1628 @@
+    std::vector<LocationIndex> targets;
+
+    LocationInfo(Location location) : location(location) {}
+  };
+
+  // Maps location indexes to the info stored there, as just described above.
+  std::vector<LocationInfo> locations;
+
+  // Reverse mapping of locations to their indexes.
+  std::unordered_map<Location, LocationIndex> locationIndexes;
+
+  const Location& getLocation(LocationIndex index) {
+    assert(index < locations.size());
+    return locations[index].location;
+  }
+
+  PossibleContents& getContents(LocationIndex index) {
+    assert(index < locations.size());
+    return locations[index].contents;
+  }
+
+private:
+  std::vector<LocationIndex>& getTargets(LocationIndex index) {
+    assert(index < locations.size());
+    return locations[index].targets;
+  }
+
+  // Convert the data into the efficient LocationIndex form we will use during
+  // the flow analysis. This method returns the index of a location, allocating
+  // one if this is the first time we see it.
+  LocationIndex getIndex(const Location& location) {
+    auto iter = locationIndexes.find(location);
+    if (iter != locationIndexes.end()) {
+      return iter->second;
+    }
+
+    // Allocate a new index here.
+    size_t index = locations.size();
+#if defined(POSSIBLE_CONTENTS_DEBUG) && POSSIBLE_CONTENTS_DEBUG >= 2
+    std::cout << "  new index " << index << " for ";
+    dump(location);
+#endif
+    if (index >= std::numeric_limits<LocationIndex>::max()) {
+      // 32 bits should be enough since each location takes at least one byte
+      // in the binary, and we don't have 4GB wasm binaries yet... do we?
+      Fatal() << "Too many locations for 32 bits";
+    }
+    locations.emplace_back(location);
+    locationIndexes[location] = index;
+
+    return index;
+  }
+
+  IndexLink getIndexes(const LocationLink& link) {
+    return {getIndex(link.from), getIndex(link.to)};
+  }
+
+  // See the comment on CollectedFuncInfo::childParents. This is the merged info
+  // from all the functions and the global scope.
+  std::unordered_map<LocationIndex, LocationIndex> childParents;
+
+  // The work remaining to do during the flow: locations that we need to flow
+  // content from, after new content reached them.
+  //
+  // Using a set here is efficient as multiple updates may arrive to a location
+  // before we get to processing it.
+  //
+  // The items here could be {location, newContents}, but it is more efficient
+  // to have already written the new contents to the main data structure. That
+  // avoids larger data here, and also, updating the contents as early as
+  // possible is helpful as anything reading them meanwhile (before we get to
+  // their work item in the queue) will see the newer value, possibly avoiding
+  // flowing an old value that would later be overwritten.
+#ifdef POSSIBLE_CONTENTS_INSERT_ORDERED
+  InsertOrderedSet<LocationIndex> workQueue;
+#else
+  std::unordered_set<LocationIndex> workQueue;
+#endif
```

Comment:
Could we move the `ifdef` out of the main code by wrapping two alternative `using WorkQueue = ...` statements instead?

Reply:
Can we have something like this up at the top where we currently have the conditional include?

```
#ifdef POSSIBLE_CONTENTS_INSERT_ORDERED
#include ...
using WorkQueue = ...
#else
using WorkQueeu = ...
#endif
```

It seems fine to define `WorkQueue` up at the top outside of any namespaces since it can get included anywhere outside of this file.

Reply:
We can, but the annoyance is that those `using WorkQueue = ..` must refer to `LocationIndex`. That is defined lower down in a double namespace. So what you are suggesting would look like this:
```cpp
namespace wasm {
namespace {

using LocationIndex = ..;

} // anonymous namespace
} // namespace wasm

#ifdef POSSIBLE_CONTENTS_INSERT_ORDERED
#include ...
using WorkQueue = InsertOrderedQueue<LocationIndex>;
#else
using WorkQueue = std::unordered_set<LocationIndex>;
#endif
```

Then we have an include after non-include code... so this just feels awkward to me, and not that helpful given this type is used in exactly one spot... But if you feel strongly I don't object.

> It seems fine to define WorkQueue up at the top outside of any namespaces since it can get included anywhere outside of this file.

Should that be "can't"? If not, why?

Reply:
Oh right, I meant "can't." I see what you mean about the annoyance now. Leaving this as-is sgtm.